js/gdm/loginDialog.js

```diff
@@ -931,6 +1079,56 @@ const LoginDialog = new Lang.Class({
         }
     },
 
+    _respondToSessionWorker: function(serviceName) {
+         this._updateSensitivity(false);
+         this._setWorking(true);
+         this._userVerifier.answerQuery(serviceName, this._promptEntry.get_text());
+    },
+
+    _performPasswordReset: function() {
+         this._updateSensitivity(false);
+         this._user.set_password_mode(AccountsService.UserPasswordMode.SET_AT_LOGIN);
+         let user = this._user;
+         this._userVerifier.cancel();
+         this._user = user;
+         this._beginVerificationForUser(user.get_user_name());
+         this._passwordResetCode = null;
+    },
+
+    _handleIncorrectPasswordResetCode: function(verifier, serviceName) {
+         this._updateSensitivity(true);
+         this._promptEntry.set_text('');
+         this._promptMessage.set_text(
+             _("Your unlock code was incorrect. Please try again.").format(this._passwordResetCode));
+
+         // Use an idle so that _holdForAnswer gets cleared first.
```

Comment:
Basically: as soon as this function returns, `this._holdForAnswer = null` will be executed, and that needs to happen before this code runs. When I wrote this code, I thought it would be less confusing to use an idle to delay execution rather than write `this._holdForAnswer = null`, which would work fine. The risk of getting rid of the idle is that it makes it possible that the assertion we discussed up above won't trigger when it should if this code is changed in the future. Also, and more importantly, it's confusing to set `this._holdForAnswer = null` any place other than right after `this._holdForAnswer.release()` is called. That's why I decided to keep the idle. But it's a tough call, and either way is fine by me.